nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-nar-utils/src/test/java/org/apache/nifi/nar/NativeLibFinderTest.java

```diff
@@ -0,0 +1,566 @@
+
+        // WHEN
+        // THEN
+        testFindLibrary(expected);
+    }
+
+    private void testFindLibrary(String expected) {
+        String actual = createTestSubjectForOS().findLibrary(NATIVE_LIB_NAME, tempDirectory.toFile());
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testFindLibraryShouldReturnLibLocation() throws Exception {
+        // GIVEN
+        File nativeLibDir = mock(File.class);
+
+        nativeLibDirs = Arrays.asList(nativeLibDir);
+
+        nativeLibNameToPath = new HashMap<>();
+
+        Path libPath = createTempFile("mocked").toAbsolutePath();
+        when(testSubjectHelper.findLibrary("libName", nativeLibDir)).thenReturn("libLocation");
+        when(testSubjectHelper.createTempCopy("libName", "libLocation")).thenReturn(libPath);
+
+        String expected = libPath.toFile().getAbsolutePath();
+
+        NativeLibFinder testSubject = createTestSubject();
+
+        // WHEN
+        String actual = testSubject.findLibrary("libName");
+
+        // THEN
+        assertEquals(expected, actual);
+        verify(testSubjectHelper).findLibrary("libName", nativeLibDir);
+        verify(testSubjectHelper).createTempCopy("libName", "libLocation");
+        verifyNoMoreInteractions(testSubjectHelper);
+    }
+
+    @Test
+    public void testFindLibraryShouldReturnFirstFoundLibLocation() throws Exception {
+        // GIVEN
+        File nativeLibDir1 = mock(File.class);
+        File nativeLibDir2 = mock(File.class);
+        File nativeLibDir3 = mock(File.class);
+
+        nativeLibDirs = Arrays.asList(nativeLibDir1, nativeLibDir2, nativeLibDir3);
+
+        nativeLibNameToPath = new HashMap<>();
+
+        Path libPath = createTempFile("mocked").toAbsolutePath();
+        when(testSubjectHelper.findLibrary("libName", nativeLibDir1)).thenReturn(null);
+        when(testSubjectHelper.findLibrary("libName", nativeLibDir2)).thenReturn("firstFoundLibLocation");
+        when(testSubjectHelper.createTempCopy("libName", "firstFoundLibLocation")).thenReturn(libPath);
+
+        String expected = libPath.toFile().getAbsolutePath();
+
+        NativeLibFinder testSubject = createTestSubject();
+
+        // WHEN
+        String actual = testSubject.findLibrary("libName");
+
+        // THEN
+        assertEquals(expected, actual);
+        verify(testSubjectHelper).findLibrary("libName", nativeLibDir1);
+        verify(testSubjectHelper).findLibrary("libName", nativeLibDir2);
+        verify(testSubjectHelper).createTempCopy("libName", "firstFoundLibLocation");
+        verifyNoMoreInteractions(testSubjectHelper);
+    }
+
+    @Test
+    public void testFindLibraryShouldReturnCachedLibLocation() throws Exception {
+        // GIVEN
+        File nativeLibDir = mock(File.class);
+
+        nativeLibDirs = Arrays.asList(nativeLibDir);
+
+        Path cachedLibPath = createTempFile("cached", "mocked").toAbsolutePath();
+        nativeLibNameToPath = new HashMap<String, Path>() {{
```

Comment:
Should avoid creating new subclass of HashMap and instead use `Collections.singletonMap`